File: 25-Dynamic-Programming/python/06_maximum_sum_increasing_subsequence.py
from typing import List

class Solution:
    def maxSumIncreasingSubsequence(self, arr: List[int], n: int) -> int:
        dp = list(arr)
        max_sum = arr[0]

        for i in range(1, n):
            for j in range(i):
                if arr[j] < arr[i] and dp[j] + arr[i] > dp[i]:
                    dp[i] = dp[j] + arr[i]
            max_sum = max(max_sum, dp[i])

        return max_sum

File: 25-Dynamic-Programming/cpp/06-maximum-sum-increasing-subsequence.cpp
#include <vector>
#include <algorithm>

using namespace std;

class Solution {
public:
    int maxSumIncreasingSubsequence(vector<int>& arr, int n) {
        vector<int> dp(n);
        for (int i = 0; i < n; i++) {
            dp[i] = arr[i];
        }

        int maxSum = arr[0];
        for (int i = 1; i < n; i++) {
            for (int j = 0; j < i; j++) {
                if (arr[j] < arr[i] && dp[j] + arr[i] > dp[i]) {
                    dp[i] = dp[j] + arr[i];
                }
            }
            maxSum = max(maxSum, dp[i]);
        }

        return maxSum;
    }
};
